lang/src/org/partiql/lang/eval/physical/operators/WindowOperatorFactoryDefault.kt

```diff
@@ -0,0 +1,111 @@
+}
+
+internal class WindowOperatorDefault(
+    private val source: RelationExpression,
+    private val windowPartitionList: List<ValueExpression>,
+    private val windowSortSpecList: List<CompiledSortKey>,
+    private val compiledWindowFunction: CompiledWindowFunction
+) : RelationExpression {
+    override fun evaluate(state: EvaluatorState): RelationIterator {
+        // the following corresponding to materialization process
+        val sourceIter = source.evaluate(state)
+        val registers = sequence {
+            while (sourceIter.nextRow()) {
+                yield(state.registers.clone())
+            }
+        }
+
+        val partitionSortSpec = windowPartitionList.map {
+            CompiledSortKey(NaturalExprValueComparators.NULLS_FIRST_ASC, it)
+        }
+
+        val sortKeys = partitionSortSpec + windowSortSpecList
+
+        val sortedRegisters = registers.sortedWith(getSortingComparator(sortKeys, state))
+
+        // create the partition here
+        var partition = mutableListOf<List<Array<ExprValue>>>()
+
+        // entire partition
+        if (windowPartitionList.isEmpty()) {
+            partition.add(sortedRegisters.toList())
+        }
+        // need to be partitioned
+        else {
+            val iter = sortedRegisters.iterator()
+            var rowInPartition = mutableListOf<Array<ExprValue>>()
+            var previousPartition: ExprValue? = null
+            while (iter.hasNext()) {
+                val currentRow = iter.next()
+                state.load(currentRow)
+                val currentPartition = state.valueFactory.newSexp(
+                    windowPartitionList.map {
+                        it.invoke(state)
+                    }
+                )
+                // for the first time,
+                if (previousPartition == null) {
+                    rowInPartition.add(currentRow)
+                    previousPartition = currentPartition
+                } else if (previousPartition.exprEquals(currentPartition)) {
+                    rowInPartition.add(currentRow)
+                } else {
+                    partition.add(rowInPartition.toList())
+                    rowInPartition.clear()
+                    previousPartition = currentPartition
+                    rowInPartition.add(currentRow)
+                }
+            }
+            // finish up
+            partition.add(rowInPartition.toList())
+            rowInPartition.clear()
+        }
+
+        return relation(RelationType.BAG) {
+            partition.forEach { rowsInPartition ->
+                val windowFunc = compiledWindowFunction.func
+                // set the window function partition to the current partition
+                windowFunc.reset(rowsInPartition)
+
+                rowsInPartition.forEach { row ->
+                    // reset state
+                    state.load(row)
```

Comment:
With the implementation of NavigationWindowFunction, I don't think this load is necessary.